File: 1-Basic technical skills for job interviews/2-Stacks, Queues, and Heaps/stack_implementation.cpp
#include <bits/stdc++.h>
using namespace std;

///Base class for Stack Operations
template<typename T>
class AbstractStack{
public:
    virtual void push(T value) = 0;
    virtual void pop() = 0;
    virtual int topElement ( ) = 0 ;
    virtual int sizeOfStack() = 0 ;
    virtual bool isEmpty() = 0 ;
    virtual void display()=0;
};

template<typename T>
class StackOperations : public AbstractStack<T> {
private:
    vector<T>elements;
    int top ;

public:

    StackOperations(){
        this->top = -1 ;
    }

    void push(T value)override{
        elements.push_back(value );
        top++;
    }

    void pop()override{
        if(top == -1 ) throw underflow_error("Stack is empty you can't remove any thing ");
        elements.pop_back();
        top -- ;

    }

    int topElement()override{
        if(top == -1 ) throw underflow_error("Stack is empty");
        int value = elements[top--];
        return value;
    }
    int sizeOfStack() override{
        return top+1;
    }

    bool isEmpty()override{
        if(top == -1 )return false;

        return true;
    }

    void display()override{
        while(top != -1 ){
            cout << elements[top--]<< " " ;
        }
        cout <<endl;
    }

};

File: 1-Basic technical skills for job interviews/2-Stacks, Queues, and Heaps/stack_implementation.dart
abstract class AbstractStack<T> {
  void push(T value);
  void pop();
  T topElement();
  int sizeOfStack();
  bool isEmpty();
  void display();
}

class StackOperations<T> implements AbstractStack<T> {
  List<T> elements = [];
  int top = -1;

  @override
  void push(T value) {
    elements.add(value);
    top++;
  }

  @override
  void pop() {
    if (top == -1) {
      throw StateError("Stack is empty. You can't remove anything.");
    }
    elements.removeLast();
    top--;
  }

  @override
  T topElement() {
    if (top == -1) {
      throw StateError("Stack is empty.");
    }
    return elements[top];
  }

  @override
  int sizeOfStack() {
    return top + 1;
  }

  @override
  bool isEmpty() {
    return top == -1;
  }

  @override
  void display() {
    while (top != -1) {
      print('${elements[top--]} ');
    }
    print('\n');
  }
}


